File: system-tests-fixtures/api/TestApi/android/lib/src/main/java/com/TestApi/ern/api/WalmartItemApi.java
/*
* Copyright 2017 WalmartLabs
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
* http://www.apache.org/licenses/LICENSE-2.0
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

package com.TestApi.ern.api;

import android.support.annotation.NonNull;

import com.walmartlabs.electrode.reactnative.bridge.ElectrodeBridgeEventListener;
import com.walmartlabs.electrode.reactnative.bridge.ElectrodeBridgeEvent;
import com.walmartlabs.electrode.reactnative.bridge.ElectrodeBridgeRequestHandler;
import com.walmartlabs.electrode.reactnative.bridge.ElectrodeBridgeResponseListener;
import com.walmartlabs.electrode.reactnative.bridge.None;
import java.util.*;
import java.util.UUID;

import com.TestApi.ern.model.Item;

public final class WalmartItemApi {
    private static final Requests REQUESTS;
    private static final Events EVENTS;

    static {
        REQUESTS = new WalmartItemRequests();
        EVENTS = new WalmartItemEvents();
    }

    private WalmartItemApi() {
    }

    @NonNull
    public static Requests requests() {
        return REQUESTS;
    }

    @NonNull
    public static Events events() {
        return EVENTS;
    }

    public interface Events {
        String EVENT_ITEM_ADDED = "com.TestApi.ern.api.event.itemAdded";

        UUID addItemAddedEventListener(@NonNull final ElectrodeBridgeEventListener<String> eventListener);

            ElectrodeBridgeEventListener<ElectrodeBridgeEvent> removeItemAddedEventListener(@NonNull final UUID uuid);

        void emitItemAdded(@NonNull String itemId);

    }

    public interface Requests {
        String REQUEST_ADD_ITEM = "com.TestApi.ern.api.request.addItem";
        String REQUEST_FIND_ITEMS = "com.TestApi.ern.api.request.findItems";


        void registerAddItemRequestHandler(@NonNull final ElectrodeBridgeRequestHandler<Item, Boolean> handler);

        void registerFindItemsRequestHandler(@NonNull final ElectrodeBridgeRequestHandler<Integer, List<Item>> handler);

        void addItem(Item item, @NonNull final ElectrodeBridgeResponseListener<Boolean> responseListener);

        void findItems(Integer limit, @NonNull final ElectrodeBridgeResponseListener<List<Item>> responseListener);

    }
}

File: system-tests/fixtures/api/TestApi/IOS/WalmartItemAPI.swift
#if swift(>=4.0)
@objcMembers public class WalmartItemAPI: NSObject  {

    static let kRequestAddItem = "com.TestApi.ern.api.request.addItem";

    static let kRequestFindItems = "com.TestApi.ern.api.request.findItems";
    static let kEventItemAdded = "com.TestApi.ern.api.event.itemAdded";


    public lazy var events: WalmartItemAPIEvents = {
        return WalmartItemEvents()
    }()


    public lazy var requests: WalmartItemAPIRequests = {
        return WalmartItemRequests()
    }()
}

@objcMembers public class WalmartItemAPIEvents: NSObject {
    public func addItemAddedEventListener(eventListener: @escaping ElectrodeBridgeEventListener) -> UUID?{
        assertionFailure("should override")
        return UUID()
    }

    public func removeItemAddedEventListener(uuid: UUID) -> ElectrodeBridgeEventListener? {
        assertionFailure("should override")
        return nil
    }

    public func emitEventItemAdded(itemId: String) {
        assertionFailure("should override")

    }
}

@objcMembers public class WalmartItemAPIRequests: NSObject {
    public func registerAddItemRequestHandler(handler: @escaping ElectrodeBridgeRequestCompletionHandler) -> UUID?{
        assertionFailure("should override")
        return UUID()
    }

    public func registerFindItemsRequestHandler(handler: @escaping ElectrodeBridgeRequestCompletionHandler) -> UUID?{
        assertionFailure("should override")
        return UUID()
    }


    public func unregisterAddItemRequestHandler(uuid: UUID) -> ElectrodeBridgeRequestCompletionHandler? {
        assertionFailure("should override")
        return nil
    }

    public func unregisterFindItemsRequestHandler(uuid: UUID) -> ElectrodeBridgeRequestCompletionHandler? {
        assertionFailure("should override")
        return nil
    }


    public func addItem(item: Item, responseCompletionHandler: @escaping ElectrodeBridgeResponseCompletionHandler) {
        assertionFailure("should override")
    }

    public func findItems(limit: Int, responseCompletionHandler: @escaping ElectrodeBridgeResponseCompletionHandler) {
        assertionFailure("should override")
    }

}
#else
public class WalmartItemAPI: NSObject  {

    static let kRequestAddItem = "com.TestApi.ern.api.request.addItem";

    static let kRequestFindItems = "com.TestApi.ern.api.request.findItems";
    static let kEventItemAdded = "com.TestApi.ern.api.event.itemAdded";


    public lazy var events: WalmartItemAPIEvents = {
        return WalmartItemEvents()
    }()


    public lazy var requests: WalmartItemAPIRequests = {
        return WalmartItemRequests()
    }()
}

public class WalmartItemAPIEvents: NSObject {
    public func addItemAddedEventListener(eventListener: @escaping ElectrodeBridgeEventListener) -> UUID?{
        assertionFailure("should override")
        return UUID()
    }

    public func removeItemAddedEventListener(uuid: UUID) -> ElectrodeBridgeEventListener? {
        assertionFailure("should override")
        return nil
    }

    public func emitEventItemAdded(itemId: String) {
        assertionFailure("should override")

    }
}

public class WalmartItemAPIRequests: NSObject {
    public func registerAddItemRequestHandler(handler: @escaping ElectrodeBridgeRequestCompletionHandler) -> UUID?{
        assertionFailure("should override")
        return UUID()
    }

    public func registerFindItemsRequestHandler(handler: @escaping ElectrodeBridgeRequestCompletionHandler) -> UUID?{
        assertionFailure("should override")
        return UUID()
    }


    public func unregisterAddItemRequestHandler(uuid: UUID) -> ElectrodeBridgeRequestCompletionHandler? {
        assertionFailure("should override")
        return nil
    }

    public func unregisterFindItemsRequestHandler(uuid: UUID) -> ElectrodeBridgeRequestCompletionHandler? {
        assertionFailure("should override")
        return nil
    }


    public func addItem(item: Item, responseCompletionHandler: @escaping ElectrodeBridgeResponseCompletionHandler) {
        assertionFailure("should override")
    }

    public func findItems(limit: Int, responseCompletionHandler: @escaping ElectrodeBridgeResponseCompletionHandler) {
        assertionFailure("should override")
    }

}

#endif
